File: cdk/python/rds-aurora-postgres/rds_aurora_postgres/rds_aurora_postgres_stack.py
from aws_cdk import (
    Duration,
    Stack,
    aws_rds as rds,
    aws_ec2 as ec2,
    aws_kms as kms,
)
from constructs import Construct

class RdsAuroraPostgresStack(Stack):

    def __init__(self, scope: Construct, construct_id: str, **kwargs) -> None:
        super().__init__(scope, construct_id, **kwargs)

        # Create basic VPC with cidr block 10.0.0.0/16
        vpc = ec2.Vpc(
            self,
            "VPC",
            ip_addresses=ec2.IpAddresses.cidr("10.0.0.0/16"),
            restrict_default_security_group=False,
            max_azs=2,
            nat_gateways=0,
            subnet_configuration=[
                ec2.SubnetConfiguration(
                    subnet_type=ec2.SubnetType.PRIVATE_ISOLATED,
                    cidr_mask=24,
                    name="RDS"
                ),
            ],
        )

        # Create username and password secret for RDS instance
        rdsusernamesecret = rds.DatabaseSecret(
            self,
            "RDSSecret",
            username="admin",
            encryption_key=kms.Key(
                self,
                "MySQL-EncryptionKey",
                enable_key_rotation=True,
                ),
        )

        # Create Aurora cluster writer instance
        aurora_writer=rds.ClusterInstance.provisioned(
            "writer",
            instance_type=ec2.InstanceType("t2.micro"),
            auto_minor_version_upgrade=True,
        )

        # Create Aurora cluster reader instance
        aurora_readers=[
            rds.ClusterInstance.provisioned(
            "reader",
            instance_type=ec2.InstanceType("t2.micro"),
            auto_minor_version_upgrade=True,
            ),
        ]

        # Create Aurora PostgreSQL cluster with version 15.3 that passes security and linting checks
        aurora_cluster = rds.DatabaseCluster(
            self,
            "AuroraCluster",
            engine=rds.DatabaseClusterEngine.aurora_postgres(version=rds.AuroraPostgresEngineVersion.VER_15_3),
            writer=aurora_writer,
            readers=aurora_readers,
            backup=rds.BackupProps(
                retention=Duration.days(1),
            ),
            credentials=rds.Credentials.from_secret(rdsusernamesecret),
            vpc=vpc,
            vpc_subnets=ec2.SubnetSelection(subnet_type=ec2.SubnetType.PRIVATE_ISOLATED),
            backtrack_window=Duration.seconds(30),
            cloudwatch_logs_exports=["postgresql"],
            monitoring_interval=Duration.seconds(5),
            iam_authentication=True,
            storage_encrypted=True,
            storage_encryption_key=kms.Key(
                self,
                "AuroraStorageEncryptionKey",
                enable_key_rotation=True,
            ),
            deletion_protection=True,
        )

        rdsusernamesecret.node.default_child.cfn_options.metadata = {
            "guard": {
                "SuppressedRules": [
                    "SECRETSMANAGER_ROTATION_ENABLED_CHECK"
                ]
            }
        }

        # Suppress Cloudformation Guard settings that do not apply to Aurora Cluster Instances, inherited from
        # the Aurora Database Cluster Properties
        aurora_cluster.node.find_child("writer").node.default_child.cfn_options.metadata = {
            "guard": {
                "SuppressedRules": [
                    "DB_INSTANCE_BACKUP_ENABLED",
                    "RDS_STORAGE_ENCRYPTED",
                    "RDS_INSTANCE_LOGGING_ENABLED",
                    "RDS_INSTANCE_DELETION_PROTECTION_ENABLED",
                    "RDS_SNAPSHOT_ENCRYPTED",
                    "RDS_MULTI_AZ_SUPPORT",
                ]
            }
        }

        # Suppress Cloudformation Guard settings that do not apply to Aurora Cluster Instances, inherited from
        # the Aurora Database Cluster Properties
        aurora_cluster.node.find_child("reader").node.default_child.cfn_options.metadata = {
            "guard": {
                "SuppressedRules": [
                    "DB_INSTANCE_BACKUP_ENABLED",
                    "RDS_STORAGE_ENCRYPTED",
                    "RDS_INSTANCE_LOGGING_ENABLED",
                    "RDS_INSTANCE_DELETION_PROTECTION_ENABLED",
                    "RDS_SNAPSHOT_ENCRYPTED",
                    "RDS_MULTI_AZ_SUPPORT",
                ]
            }
        }

File: cdk/java/rds-aurora-postgres/src/main/java/com/myorg/RdsAuroraPostgresStack.java
package com.myorg;

import software.constructs.Construct;

import java.util.List;

import software.amazon.awscdk.Stack;
import software.amazon.awscdk.StackProps;
import software.amazon.awscdk.Duration;
import software.amazon.awscdk.services.kms.*;
import software.amazon.awscdk.services.ec2.Vpc;
import software.amazon.awscdk.services.ec2.IpAddresses;
import software.amazon.awscdk.services.ec2.SubnetType;
import software.amazon.awscdk.services.ec2.SubnetConfiguration;
import software.amazon.awscdk.services.ec2.InstanceClass;
import software.amazon.awscdk.services.ec2.InstanceSize;
import software.amazon.awscdk.services.ec2.InstanceType;
import software.amazon.awscdk.services.ec2.SubnetSelection;
import software.amazon.awscdk.services.rds.*;
import software.amazon.awscdk.services.rds.AuroraPostgresEngineVersion;
import software.amazon.awscdk.services.rds.DatabaseInstance;
import software.amazon.awscdk.services.rds.DatabaseSecret;
import software.amazon.awscdk.services.rds.DatabaseInstanceEngine;
import software.amazon.awscdk.services.rds.Credentials;
import software.amazon.awscdk.services.rds.DatabaseClusterEngine;
import software.amazon.awscdk.services.rds.ClusterInstance;
//import software.amazon.awscdk.services.rds.InstanceType;
// import software.amazon.awscdk.services.sqs.Queue;

public class RdsAuroraPostgresStack extends Stack {
    public RdsAuroraPostgresStack(final Construct scope, final String id) {
        this(scope, id, null);
    }

    public RdsAuroraPostgresStack(final Construct scope, final String id, final StackProps props) {
        super(scope, id, props);

    
    //Create basic VPC with cidr block 10.0.0.0/16
         Vpc vpc = Vpc.Builder.create(this, "vpc-cw-iac-example")
         .ipAddresses(IpAddresses.cidr("10.0.0.0/16"))
         .maxAzs(2)
         .restrictDefaultSecurityGroup(false)
         .natGateways(0)
         .subnetConfiguration(List.of(SubnetConfiguration.builder()
                                .name("RDS")
                                .subnetType(SubnetType.PRIVATE_ISOLATED)
                                  // the properties below are optional
                                .cidrMask(24)
                                .build()))
         .build();

    //Create username and password secret for RDS instance
        DatabaseSecret rdsusernamesecret = DatabaseSecret.Builder.create(this, "rds-secret-cw-iac-example")
        .username("clusteradmin")
        .encryptionKey(Key.Builder.create(this, "kms-secret-cw-iac-example")
                         .enableKeyRotation(true)
                       .build())
        .build();

    //Create Aurora PostgreSQL cluster with version 15. 
        DatabaseCluster aurorainstance = DatabaseCluster.Builder.create(this, "rds-aurora-postgres-cw-iac-example")
        .engine(DatabaseClusterEngine.auroraPostgres(AuroraPostgresClusterEngineProps.builder().version(AuroraPostgresEngineVersion.VER_15_2).build()))
        .writer(ClusterInstance.provisioned("writer", ProvisionedClusterInstanceProps.builder()
                .instanceType(InstanceType.of(InstanceClass.R6G, InstanceSize.XLARGE4))
                .build()))
        .serverlessV2MinCapacity(6.5)
        .serverlessV2MaxCapacity(64)
        .readers(List.of(ClusterInstance.serverlessV2("reader1", ServerlessV2ClusterInstanceProps.builder().scaleWithWriter(true).build()), ClusterInstance.serverlessV2("reader2")))
        .vpc(vpc)
        .vpcSubnets(SubnetSelection.builder().subnetType(SubnetType.PRIVATE_ISOLATED).build())
        .deletionProtection(true)
        .storageEncrypted(true)
        .storageEncryptionKey(Key.Builder.create(this, "aurora-encryptionkey-cw-iac-example").enableKeyRotation(true).build())
        .iamAuthentication(true)
        .monitoringInterval(Duration.seconds(0))
        .credentials(Credentials.fromSecret(rdsusernamesecret))
        .cloudwatchLogsExports(List.of("postgresql"))
        .backup(BackupProps.builder().retention(Duration.days(1)).build())
        .build();

    }     
}
